### PR TITLE
14 - Moving from pulling policies from rego to pushing them from Anubis Policy API

### PR DESCRIPTION
**Pull request by Cerfoglg**
## Proposed changes

Instead of pulling all policies from OPA on every call being processed, we instead push the policies from the Policy API when a policy is created/edited/deleted

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [Contributor License Agreement][cla]
- [x] I have updated the [RELEASE NOTES][release]
- [x] I have added tests that prove my fix is effective or that my
      feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/anubis/master/individual_cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/orchestracities/anubis/blob/master/CONTRIBUTING.md
    "Contributing to Anubis"
[release]: https://github.com/orchestracities/anubis/blob/master/RELEASE_NOTES.md
    "Anubis Release Notes"

### CODE REVIEW

File: config/opa-service/rego/common.rego [hunk 1]
```diff
@@ -54,8 +54,11 @@ jwks_request(url) = http.send({
     "force_cache_duration_seconds": 3600
 })
 
-# Grab data from API
-data = http.send({"method": "get", "url": api_uri, "headers": {"accept": "text/rego", "fiware-service": request.tenant, "fiware-servicepath": request.service_path}}).body
+# Policies pull method
+# policies = http.send({"method": "get", "url": api_uri, "headers": {"accept": "text/rego", "fiware-service": request.tenant, "fiware-servicepath": request.service_path}}).body
```
Comment (chicco785) [inline config/opa-service/rego/common.rego:58]: we should document how to enable set pull or push method, if possible the best solution would be to have an environment variable to pick one or the other method.

File: docker-compose.yaml [hunk 2]
```diff
@@ -35,6 +35,7 @@ services:
       - KEYCLOACK_ENABLED=True
       - TENANT_ADMIN_ROLE_ID=9dc79aa8-d42f-4720-8de0-fe79a00a46b7
       - KEYCLOACK_ADMIN_ENDPOINT=http://keycloak:8080/admin/realms/default
+      - OPA_ENDPOINT=http://ext_authz-opa-service:8181
```
Comment (chicco785) [inline docker-compose.yaml:38]: this needs to be documented

File: docker-compose.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ services:
       - ./config/envoy/v3.yaml:/etc/envoy/envoy.yaml
       - ./config/opa-service/lua:/etc/envoy/lua
     environment:
-      - loglevel=error
+      - loglevel=debug
```
Comment (chicco785) [inline docker-compose.yaml:10]: let's keep error to merge

File: docker-compose.yaml [hunk 4]
```diff
@@ -58,7 +59,7 @@ services:
       - ./config/opa-service/opa.yaml:/opa.yaml
     command:
       - run
-      - --log-level=error
+      - --log-level=info
```
Comment (chicco785) [inline docker-compose.yaml:62]: let's keep error to merge

File: anubis-management-api/anubis/policies/routers.py [hunk 5]
```diff
@@ -508,6 +509,23 @@ def create_policy(
                 urllib.parse.quote(
                     policy.access_to)))
         logging.error(e)
+    opa_url = os.environ.get('OPA_ENDPOINT', 'http://127.0.0.1:8181')
```
Comment (chicco785) [inline anubis-management-api/anubis/policies/routers.py:512]: code below should be executed only if OPA_ENDPOINT is specified (leaving a way to use pull mode)

File: anubis-management-api/anubis/policies/routers.py [hunk 6]
```diff
@@ -581,6 +599,23 @@ def update(
...
```
Comment (chicco785) [inline anubis-management-api/anubis/policies/routers.py:602]: code below should be executed only if OPA_ENDPOINT is specified (leaving a way to use pull mode)

File: anubis-management-api/anubis/policies/routers.py [hunk 7]
```diff
@@ -631,4 +666,21 @@ def delete_policy(
                 urllib.parse.quote(
                     db_policy.access_to)))
         logging.error(e)
+    opa_url = os.environ.get('OPA_ENDPOINT', 'http://127.0.0.1:8181')
```
Comment (chicco785) [inline anubis-management-api/anubis/policies/routers.py:669]: code below should be executed only if OPA_ENDPOINT is specified (leaving a way to use pull mode)

**Review by chicco785 — CHANGES_REQUESTED** [review]
see comments from my side, also when the api starts, it should push policies, otherwise this are pushed to opa only when a change occurs. we should, in a future interaction also understand that happens when opa restart for whatever reason (is there persistency?)

**Comment by chicco785** [comment]
> see comments from my side, also when the api starts, it should push policies, otherwise this are pushed to opa only when a change occurs. we should, in a future interaction also understand that happens when opa restart for whatever reason (is there persistency?)

alternatively, we should have a "sidecar" to opa that read data from the api on start. may be better as an overall option and solves the issue also of opa restarts without need for persistency.

File: docker-compose.yaml (hunk 2 above)
Comment (chicco785) [reply docker-compose.yaml:38]: also, considering there may be for multiple instances of opa deployment, comma separation support would be nice.

File: config/opa-service/rego/common.rego (hunk 1 above)
Comment (Cerfoglg) [reply config/opa-service/rego/common.rego:58]: This commented code was only left for the sake of keeping it "in mind". There's really no reason to use it when the push approach is better anyways. Maybe we can think of a different push approach later on.

File: scripts/init_opa.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh -e
+
+apk update
+apk add curl
+
+curl -s -i -X 'GET' \
```
Comment (chicco785) [inline scripts/init_opa.sh:6]: can't we iterate over the api and extract the list of tenants from which policies needs to be extracted?

**Comment by chicco785** [comment]
hi @Cerfoglg testing, i got a few errors:

e.g., when deleting a policy:
```
anubis-policy-api-1             | Traceback (most recent call last):
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 404, in run_asgi
anubis-policy-api-1             |     result = await app(  # type: ignore[func-returns-value]
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
anubis-policy-api-1             |     return await self.app(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
anubis-policy-api-1             |     await super().__call__(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
anubis-policy-api-1             |     await self.middleware_stack(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
anubis-policy-api-1             |     raise exc
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
anubis-policy-api-1             |     await self.app(scope, receive, _send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 92, in __call__
anubis-policy-api-1             |     await self.simple_response(scope, receive, send, request_headers=headers)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 147, in simple_response
anubis-policy-api-1             |     await self.app(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 93, in __call__
anubis-policy-api-1             |     raise exc
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
anubis-policy-api-1             |     await self.app(scope, receive, sender)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
anubis-policy-api-1             |     raise e
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
anubis-policy-api-1             |     await self.app(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/routing.py", line 670, in __call__
anubis-policy-api-1             |     await route.handle(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/routing.py", line 266, in handle
anubis-policy-api-1             |     await self.app(scope, receive, send)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/routing.py", line 65, in app
anubis-policy-api-1             |     response = await func(request)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/fastapi/routing.py", line 227, in app
anubis-policy-api-1             |     raw_response = await run_endpoint_function(
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
anubis-policy-api-1             |     return await run_in_threadpool(dependant.call, **values)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
anubis-policy-api-1             |     return await anyio.to_thread.run_sync(func, *args)
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync
anubis-policy-api-1             |     return await get_asynclib().run_sync_in_worker_thread(
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
anubis-policy-api-1             |     return await future
anubis-policy-api-1             |   File "/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
anubis-policy-api-1             |     result = context.run(func, *args)
anubis-policy-api-1             |   File "/home/apiuser/anubis-management-api/./anubis/policies/routers.py", line 676, in delete_policy
anubis-policy-api-1             |     service_path_id=db_service_path_id)
anubis-policy-api-1             | NameError: name 'db_service_path_id' is not defined
anubis-policy-api-1             | time=2023-01-12T17:07:28.442Z  | lvl=INFO:     | comp=uvicorn.access | msg="172.20.0.5:51416 - "POST /v1/audit/logs HTTP/1.1" 200
anubis-mongo-1                  | {"t":{"$date":"2023-01-12T17:07:28.545+00:00"},"s":"I",  "c":"STORAGE",  "id":22430,   "ctx":"WTCheckpointThread","msg":"WiredTige
```

when running the load script:
```
anubis-opa_init_script-1        | Content-Length: 103
anubis-opa_init_script-1        | 
anubis-opa_init_script-1        | {
anubis-opa_init_script-1        |   "code": "invalid_parameter",
anubis-opa_init_script-1        |   "message": "invalid character 'H' looking for beginning of value"
anubis-opa_init_script-1        | }
```

when running the context broker tests:
```
Can I create an entity in ServicePath / for Tenant1?
===============================================================
ERROR: Can't create entity
```


that's why unit test in python should also use OPA (and not only the e2e).

File: anubis-management-api/anubis/policies/routers.py [hunk 9]
```diff
@@ -508,6 +509,24 @@ def create_policy(
                 urllib.parse.quote(
                     policy.access_to)))
         logging.error(e)
+    if os.environ.get('OPA_ENDPOINT'):
+        opa_url = os.environ.get('OPA_ENDPOINT', 'http://127.0.0.1:8181')
+        db_policies = operations.get_policies_by_service_path(
+            db,
+            tenant=fiware_service,
+            service_path_id=db_service_path_id)
```
Comment (c0c0n3) [inline anubis-management-api/anubis/policies/routers.py:517]: I'm assuming this will work even if there's no service path? think Teadal...

**Comment by c0c0n3** [comment]
Great performance improvement, well done! and yea, maybe mention/update the figures it in the docs?

also, re: performance. one thing to look out for is to avoid a situation where you push a huge amount of records from the db to opa---e.g. control policy we talked about earlier in the mtg...

**Comment by c0c0n3** [comment]
just a side note. as we talked about in the past, the general assumption when replicating data in anubis is that updates are infrequent and almost never happen concurrently.

why am I bringing this up again? well, I could be wrong, but if I understand what's happening under the bonnet (and it's a big if), then policies data replicated in OPA could be inconsistent if we bring concurrent writes into the mix. Here's the scenario

- Anubis holds policies for tenant Tenny: `[p1, p2]`.
- `p1` is Bob's policy whereas `p2` is Alice's.
- Alice writes a new policy `p3`.
- At the same time, Bob deletes `p1`.

What will end up in OPA? It's possible OPA eventually contains `[p2]` (or even `[p1, p2, p3]`) instead of `[p2, p3]` if I interpret the code correctly. In fact, here's a possible execution history:

1. (Bob's tread) Anubis deletes `p1` and then reads all of Tenny's policies, so it gets `[p2]`.
2. (Alice's thread) Anubis writes `p3` and reads all of Tenny's policies, so it gets `[p2, p3]`.
3. (Alice's thread) Anubis `PUT`s `[p2, p3]` into OPA.
4. (Bob's tread) Anubis `PUT`s `[p2]` into OPA.

File: anubis-management-api/anubis/policies/routers.py (hunk 9 above)
Comment (c0c0n3) [reply anubis-management-api/anubis/policies/routers.py:517]: also, it looks like `get_policies_by_service_path` comes w/ a default limit param set to 100? what happens if there's more than 100 policies under that tenant/path? again, think Teadal...

Comment (chicco785) [reply anubis-management-api/anubis/policies/routers.py:517]: for now put `-1` as limit

**Review by chicco785 — CHANGES_REQUESTED** [review]
@Cerfoglg see pending comments.
Also `README.md` was not up-to-date, I made few changes, but better have a look.

File: scripts/test_context_broker.sh [hunk 10]
```diff
@@ -181,4 +181,15 @@ else
...
+echo "==============================================================="
+export response=`curl -s -o /dev/null -w "%{http_code}" 'http://localhost:8181/v1/data/Tenant2/policies'`
+if [ $response == "200" ]
+then
+  echo "PASSED"
+else
+  echo "ERROR: Policies haven't been pushed to OPA"
+  exit 1
+fi
+
```
[the remark this answers is not in the document]
Comment (Cerfoglg) [reply scripts/test_context_broker.sh:205]: It's really messy to do with just curl and grep in bash. I think what's best to do is one pull where the tests are reworked to use jq, so you can better check the JSON responses in a cleaner manner for these tests and future ones I would say.

[the remark this answers is not in the document]
Comment (chicco785) [reply scripts/test_context_broker.sh:205]: agree that testing with jq is fine

File: anubis-management-api/anubis/tenants/routers.py [hunk 11]
```diff
@@ -162,6 +164,29 @@ def create_tenant(
 
     tenant_id = operations.create_tenant(
         db=db, tenant=tenant, tenant_id=db_tenant_id).id
+
+    if os.environ.get('OPA_ENDPOINT'):
```
Comment (chicco785) [inline anubis-management-api/anubis/tenants/routers.py:168]: is this any different from `update_opa_policies` ?

**Review by chicco785 — APPROVED** [review]

**Review by c0c0n3 — APPROVED** [review]

**Pull request merged**